src/biconnectivity13/biconnect.jl

```diff
@@ -84,16 +27,72 @@ julia> biconnected_components(cycle_graph(5))
 """
 function biconnected_components end
 @traitfn function biconnected_components(g::::(!IsDirected))
-    state = Biconnections(g)
-    for u in vertices(g)
-        if state.depth[u] == 0
-            visit!(g, state, u, u)
-        end
+        n = nv(g)
+        S = Array{Tuple{Int, Int}, 1}(undef,n)
+        E = Edge{eltype(g)}
+        edge_st = Vector{E}()
+        components = Vector{Vector{E}}()
+        idx = zero(Int)      # preorder counter
+        low = zeros(Int, nv(g))
+        pre = zeros(Int, nv(g))
+        #stack pointer to emulate what happens in function call,
+        #that will make us avoid the reallocation of push and pop vector
+        stack_ptr = zero(Int) #stack pointer to emulate what happens in function  call, that will make us avoid the reallocation of push and pop vector
+
+    @inbounds for us in vertices(g)
+        pre[us] == 0 || continue # don’t go to visited nodes again
+        stack_ptr += 1
+        S[stack_ptr] = (us, 1)
+        ptr = one(Int)
+
+        while stack_ptr > 0
+            v, _ = S[stack_ptr]
```

Comment:
since `i`, below, is `eltype(g)`, `v` is also `eltype(g)`, which means the first element in the tuple definition for `S` is incorrect.